src/vm.rs

```diff
@@ -690,14 +690,63 @@ impl<T: GuestStagePageTable> Vm<T, VmStateFinalized> {
                     event_data,
                 );
                 if let Ok(counter_index) = result && counter_index < pmu_info.get_num_counters() {
-                    // TODO: Update internal state for configured counters.
+                    update_configured_counter(counter_index, config_flags, event_type, active_cpu);
                 }
                 result
             } else {
                 Err(Error::InvalidParam)
             }
         }
 
+        pub fn update_configured_counter(
+            counter_index: u64,
+            config_flags: PmuCounterConfigFlags,
+            event_type: PmuEventType,
+            active_cpu: &mut VmCpu,
+        ) {
+            let pmu_info = PmuInfo::get();
+            if pmu_info.get_counter_info(counter_index).is_some() {
+                active_cpu.set_counter_configuration(counter_index, event_type.raw());
+                if config_flags.get_auto_start() {
+                    active_cpu.set_started_counter(counter_index);
+                }
+            }
+        }
+
+        fn update_started_counters(
+            counter_index_base: u64,
+            counter_mask: u64,
+            active_cpu: &mut VmCpu,
+        ) {
+            let pmu_info = PmuInfo::get();
+            let num_counters = pmu_info.get_num_counters();
+            let configured_counters = active_cpu.get_configured_counters();
+            for i in counter_index_base..num_counters {
+                if configured_counters & (1 << i) != 0 && (i & counter_mask) != 0 {
```

Comment:
If I understand this correctly, update_started_counters is called after start_counter returns success. If the counter is not configured, the call will never succeed. 
In fact this check should be there in start counter before making the SBI call to prevent the call at the first place.

Is there a use case I am missing ?

Reply:
Got it...in that case, the check for configured counters is redundant....will remove it.

Reply:
Actually, this is scanning the configured counters to determine which counters should be marked as started. Basically, since we can start multiple configured counters based on the mask, the code is scanning the bit-mask to see what counters could have been started. Does that make sense?

Reply:
That check should happen before starting the counter not after.

Reply:
Since we are the middle layer, we can't tell whether the call will actually succeed. So, it seems premature to check for configured before the call, since the upper layer will presumably fail the call if the configuration check doesn't pass?

Reply:
This is update_started_counters function which is called after the start_call returns. start_call should not be invoked before the config_match. All the SBI calls are synchronous. Thus, a lower privilege mode must not invoke start_counter if config is not complete. So you can return the error right away. 

Any counter that is configured can only be started. The middle layer should also be aware of the configured counters before it receives the counter_start call.